File: c-sharp/Problems/arrays-and-hashing/ProductOfArrayExceptSelf.cs
/**
 * 238
 * Product of Array Except Self
 **
 * Given an integer array nums,
 * return an array answer such that
 * answer[i] is equal to the product of all the elements of nums except nums[i].
 *
 * The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 * You must write an algorithm that runs in O(n) time and without using the division operation.
 *
 * Example 1:
 *   Input: nums = [1,2,3,4]
 *   Output: [24,12,8,6]
 *
 * Example 2:
 *   Input: nums = [-1,1,0,-3,3]
 *   Output: [0,0,9,0,0]
 *
 * Constraints:
 *   • 2 <= nums.length <= 10^5
 *   • -30 <= nums[i] <= 30
 *   • The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 *
 * Follow up:
 *   Can you solve the problem in O(1) extra space complexity?
 *   (The output array does not count as extra space for space complexity analysis.)
 **
 * https://leetcode.com/problems/product-of-array-except-self/
***/

using System;

namespace Problems;

public class ProductOfArrayExceptSelf
{
    public int[] ProductExceptSelf( int[] nums )
    {
        int[] result = new int[nums.Length];

        result[0] = 1;

        for ( int i = 0; i < nums.Length - 1; i++ )
        {
            result[i + 1] = nums[i] * result[i];
        }

        int postfix = 1;

        for ( int i = nums.Length - 1; i >= 0; i-- )
        {
            result[i] *= postfix;
            postfix *= nums[i];
        }

        return result;
    }

    public int[] ProductExceptSelfCombination( int[] nums )
    {
        int size = nums.Length;

        int[] result = new int[size];
        Array.Fill( result, 1 );

        int prefix = 1;
        int postfix = 1;

        for ( int i = 0; i < size; i++ )
        {
            result[i] *= prefix;
            prefix *= nums[i];

            result[size - i - 1] *= postfix;
            postfix *= nums[size - i - 1];
        }

        return result;
    }
}

File: cpp/leetcode-cpp/product-of-array-except-self.cpp
#include <vector>
#include <iostream>

using namespace std;


/*
    Given an integer array nums, return an array such that:
    answer[i] is equal to the product of all elements of nums except nums[i]
    Ex. nums = [1,2,3,4] -> [24,12,8,6], nums = [-1,1,0,-3,3] -> [0,0,9,0,0]

    Calculate prefix products forward, then postfix backwards in a 2nd pass

    Time: O(n)
    Space: O(1)
*/

static class Solution
{
    public:
    static vector<int> productExceptSelf( vector<int>& nums )
    {
        vector<int> result;

        for ( int i = 0; i < nums.size(); i++ )
        {
            int subResult = 1;

            vector<int> subArray = nums;
            subArray.erase( subArray.begin() + i );

            for ( int j = 0; j < subArray.size(); j++ )
            {
                subResult *= subArray[j];
            }

            result.push_back( subResult );
        }

        return result;
    }

    public:
    static vector<int> productExceptSelfF( vector<int>& nums )
    {
        int inputSize = nums.size();

        vector<int> result( inputSize, 1 );

        int prefix = 1;
        for ( int i = 0; i < inputSize; i++ )
        {
            result[i] = prefix;
            prefix *= nums[i];
        }

        int postfix = 1;
        for ( int i = inputSize - 1; i >= 0; i-- )
        {
            result[i] *= postfix;
            postfix *= nums[i];
        }

        return result;
    }
};

int product_except_self_main()
{
    vector<vector<int>> input =
    {
        { 1, 2, 3, 4 },      // [24,12,8,6]
        { -1, 1, 0, -3, 3 }  // [0,0,9,0,0]
    };

    for ( int i = 0; i < input.size(); i++ )
    {
        vector<int> result = Solution::productExceptSelf( input[i] );

        for ( int j = 0; j < result.size(); j++ )
        {
            cout << result[j] << " ";
        }

        cout << "\r\n";
    }

    return 0;
}
